File: packages/js-core/src/client-internal/pairing-session-manager.ts
import { sha256 } from '@noble/hashes/sha2.js';
import { randomBytes } from '@noble/hashes/utils.js';
import type { KeyPair, PolycentricClient } from '../polycentric-client';
import * as Proto from '../proto/v2';
import { IdentityManager } from './identity-manager';

/**
 * Default expiration time for a pairing session.
 */
export const PAIRING_SESSION_TTL_MILLIS = 5 * 60 * 1000;

/** Nonce length, in bytes, for a new session digest. */
const NONCE_BYTES = 32;

/**
 * Relevant state regarding an active pairing session.
 */
export interface PairingSession {
  /** The canonical serialized `PairingSessionDigest` for this session. */
  digestBytes: Uint8Array;

  /** The decoded digest. */
  digest: Required<Proto.PairingSessionDigest>;

  /** Pairing session state verified to be from the issuer. */
  issuerState: Proto.IssuerPairingState;

  /** Public keys to that are requesting to be paired. */
  claimers: Proto.PublicKey[];

  /** The expiration time of the pairing session. */
  expiresAt: Date;

  /** The minimal information needed to access this pairing session. */
  pairingInfo: Proto.PairingInfo;
}

/**
 * Wrappers over the rs-core's pairing API.
 */
export class PairingSessionManager {
  constructor(private readonly client: PolycentricClient) {}

  /** Sign a serialized `IssuerPairingState`. */
  private async signIssuerState(
    issuerState: Proto.IssuerPairingState,
  ): Promise<Proto.SignedIssuerState> {
    const keyPair = this.requireKeyPair();

    const stateBytes = Proto.IssuerPairingState.toBinary(issuerState);
    const signature = await this.client.crypto.sign(
      keyPair.privateKey.key,
      stateBytes,
      keyPair.keyType,
    );

    return Proto.SignedIssuerState.create({ stateBytes, signature });
  }

  /**
   * Sign and publish a new pairing session state derived from local stores.
   * Returns the server's reported pairing session state or errors if the
   * response doesn't reflect the state we pushed.
   */
  private async putState(
    server: string,
    digestBytes: Uint8Array,
    sequence: bigint,
  ): Promise<PairingSession> {
    const { issuerIdentity } = this.digestFrom(digestBytes, true);

    // Get latest identity state to push
    const chain = this.client.resolveIdentityChain(issuerIdentity);
    const identityState = chain[chain.length - 1];
    if (!identityState) {
      throw new Error(`No local identity chain for ${issuerIdentity}`);
    }

    // Assemble new state
    const issuerState = Proto.IssuerPairingState.create({
      sessionDigest: digestBytes,
      identityState,
      sequence,
    });

    const signedState = await this.signIssuerState(issuerState);

    // Push the new state and handle response
    const responseBytes = await this.client.core.putPairingSession(
      server,
      Proto.SignedIssuerState.toBinary(signedState).buffer as ArrayBuffer,
    );

    return this.decodeSession(server, responseBytes);
  }

  /**
   * Start a new pairing session and register it on `server`.
   */
  async createPairingSession(server: string): Promise<PairingSession> {
    const digestBytes = Proto.PairingSessionDigest.toBinary(
      Proto.PairingSessionDigest.create({
        issuerIdentity: this.requireIdentityKey(),
        issuerSigner: this.requireKeyPair().publicKey,
        nonce: randomBytes(NONCE_BYTES),
        initialTimestamp: BigInt(Date.now()),
        ttlMillis: BigInt(PAIRING_SESSION_TTL_MILLIS),
      }),
    );

    return await this.putState(server, digestBytes, 1n);
  }

  /**
   * Publish a new state for an existing session.
   */
  async updatePairingSession(
    server: string,
    digestBytes: Uint8Array,
    sequence: bigint,
  ): Promise<PairingSession> {
    return await this.putState(server, digestBytes, sequence);
  }

  /** Fetch a pairing session using `info`. */
  async getPairingSession(info: Proto.PairingInfo): Promise<PairingSession> {
    const stateBytes = await this.client.core.getPairingSession(
      info.server,
      toCoreBytes(info.digestSha256),
    );
    return this.decodeSession(info.server, stateBytes);
  }

  /** Register our key as a claimer on a session. */
  async joinPairingSession(info: Proto.PairingInfo): Promise<void> {
    await this.client.core.joinPairingSession(
      info.server,
      toCoreBytes(info.digestSha256),
      toCoreKey(this.requireKeyPair().publicKey),
    );
  }

  /**
   * Issuer-side poll: retrieve the server's list of claimer public keys.
   */
  async pollForClaimers(info: Proto.PairingInfo): Promise<Proto.PublicKey[]> {
    const claimers = await this.client.core.pollForClaimers(
      info.server,
      toCoreBytes(info.digestSha256),
    );
    return claimers.map(fromCoreKey);
  }

  /**
   * Claimer-side poll: check if the issuer has included our key in its latest
   * identity state.
   * We should still verify the full identity chain before fully committing.
   */
  async pollForAuthorization(info: Proto.PairingInfo): Promise<boolean> {
    return await this.client.core.pollForAuthorization(
      info.server,
      toCoreBytes(info.digestSha256),
      toCoreKey(this.requireKeyPair().publicKey),
    );
  }

  /** Get the current key pair or throw. */
  private requireKeyPair(): KeyPair {
    if (!this.client.currentKeyPair) throw new Error('No active key pair');
    return this.client.currentKeyPair;
  }

  /** Get the active identity key or throw. */
  private requireIdentityKey(): string {
    if (!this.client.activeIdentityKey) throw new Error('No active identity');
    return this.client.activeIdentityKey;
  }

  /**
   * Decode a `PairingSessionState` message received from a server.
   * Use `assertIssuer === true` when updating a state to ensure that the
   * polycentric client's identity and keypair match the pairing session's.
   */
  private digestFrom(
    digestBytes: Uint8Array,
    assertIssuer = false,
  ): Required<Proto.PairingSessionDigest> {
    const decoded = Proto.PairingSessionDigest.fromBinary(digestBytes);

    const { issuerSigner } = decoded;
    if (!issuerSigner) {
      throw new Error('Pairing session digest has no issuer signer');
    }

    // All fields are known to be present in this copy.
    const digest = { ...decoded, issuerSigner };

    if (assertIssuer) {
      if (digest.issuerIdentity !== this.requireIdentityKey()) {
        throw new Error(
          `Pairing session specifies a different identity than the active identity key`,
        );
      }

      const signerMatches = IdentityManager.keysEqual(
        issuerSigner,
        this.requireKeyPair().publicKey,
      );

      if (!signerMatches) {
        throw new Error(
          'Pairing session pins a different signer than the active key pair',
        );
      }
    }

    return digest;
  }

  /**
   * Decode a `PairingSessionState` message received from a server.
   */
  private decodeSession(
    server: string,
    stateBytes: ArrayBuffer,
  ): PairingSession {
    const state = Proto.PairingSessionState.fromBinary(
      new Uint8Array(stateBytes),
    );

    const issuerState = state.issuerState;
    if (!issuerState) {
      throw new Error('Pairing session state has no signed issuer state');
    }

    const decoded = Proto.IssuerPairingState.fromBinary(issuerState.stateBytes);
    const digestBytes = decoded.sessionDigest.slice();
    const digest = this.digestFrom(digestBytes);

    return {
      digestBytes,
      digest,
      issuerState: decoded,
      claimers: state.claimers,
      expiresAt: expiresAt(digest),
      pairingInfo: Proto.PairingInfo.create({
        server,
        digestSha256: sha256(digestBytes),
      }),
    };
  }
}

/**
 * Derive the expiration time for a pairing session.
 */
function expiresAt(digest: Proto.PairingSessionDigest): Date {
  const millis = digest.initialTimestamp + digest.ttlMillis;
  const expiration = new Date(Number(millis));

  if (Number.isNaN(expiration.getTime())) {
    throw new Error(`Pairing session expiration out of range: ${millis}`);
  }

  return expiration;
}

/** Copy bytes for the FFI. */
function toCoreBytes(bytes: Uint8Array): ArrayBuffer {
  return bytes.slice().buffer as ArrayBuffer;
}

/** Convert to the FFI key type. */
function toCoreKey(key: Proto.PublicKey): {
  keyType: number;
  key: ArrayBuffer;
} {
  return {
    keyType: key.keyType,
    key: toCoreBytes(key.key),
  };
}

/** Convert back from the FFI key type. */
function fromCoreKey(key: {
  keyType: number;
  key: ArrayBuffer;
}): Proto.PublicKey {
  return Proto.PublicKey.create({
    keyType: key.keyType,
    key: new Uint8Array(key.key),
  });
}

File: packages/kt-core/core/src/main/kotlin/PairingSessionManager.kt
package org.futo.polycentric.core

import java.security.MessageDigest
import java.security.SecureRandom
import okio.ByteString
import okio.ByteString.Companion.toByteString
import org.futo.polycentric.ffi.PublicKey as FfiPublicKey
import polycentric.v2.IssuerPairingState
import polycentric.v2.KeyType
import polycentric.v2.PairingInfo
import polycentric.v2.PairingSessionDigest
import polycentric.v2.PairingSessionState
import polycentric.v2.PublicKey
import polycentric.v2.SignedIssuerState

const val PAIRING_SESSION_TTL_MILLIS = 5 * 60 * 1000L
private const val NONCE_BYTES = 32

/**
 * Device-pairing handshake manager. The pairing session creator publishes a new pairing session to
 * a server, and shares pairing info using a QR code or typable pairing code. After another device
 * joins, the user verifies each device displays the same emoji fingerprint for the pairing session
 * to confirm that no attacker has intercepted the pairing session.
 */
class PairingSessionManager(private val client: PolycentricClient) {

    class PairingSession(
        /** Canonical serialized `PairingSessionDigest` for this session. */
        val digestBytes: ByteArray,
        val digest: PairingSessionDigest,
        /** Pairing session state verified to be from the issuer. */
        val issuerState: IssuerPairingState,
        /** Public keys requesting to be paired. */
        val claimers: List<PublicKey>,
        /** Expiration time in posix epoch millis. */
        val expiresAt: Long,
        /** Information needed to access this session from the server. */
        val pairingInfo: PairingInfo,
    )

    /** Start a new pairing session and register it on [server]. */
    suspend fun createPairingSession(server: String): PairingSession {
        val keyPair = client.currentKeyPair ?: throw NoActiveKeyPairException()
        val identityKey = client.activeIdentityKey ?: throw NoActiveIdentityException()
        val digestBytes = PairingSessionDigest.ADAPTER.encode(
            PairingSessionDigest(
                issuer_identity = identityKey,
                issuer_signer = keyPair.toPublicKeyProto(),
                nonce = randomNonce(),
                initial_timestamp = System.currentTimeMillis(),
                ttl_millis = PAIRING_SESSION_TTL_MILLIS,
            ),
        )
        return putState(server, digestBytes, sequence = 1L)
    }

    /** Publish a new state for an existing session. */
    suspend fun updatePairingSession(
        server: String,
        digestBytes: ByteArray,
        sequence: Long,
    ): PairingSession = putState(server, digestBytes, sequence)

    /** Fetch a pairing session using [info]. */
    suspend fun getPairingSession(info: PairingInfo): PairingSession =
        decodeSession(
            info.server,
            client.core.getPairingSession(info.server, info.digest_sha256.toByteArray()),
        )

    /** Register our key as a claimer on a session. Verify the session first. */
    suspend fun joinPairingSession(info: PairingInfo) {
        val keyPair = client.currentKeyPair ?: throw NoActiveKeyPairException()
        client.core.joinPairingSession(
            info.server,
            info.digest_sha256.toByteArray(),
            keyPair.toPublicKeyProto().toFfi(),
        )
    }

    /** Poll the server's list of claimer public keys. */
    suspend fun pollForClaimers(info: PairingInfo): List<PublicKey> =
        client.core
            .pollForClaimers(info.server, info.digest_sha256.toByteArray())
            .map { it.toProto() }

    /**
     * Check whether the issuer's published identity state authorizes our key.
		 * Verify the full identity chain before committing.
     */
    suspend fun pollForAuthorization(info: PairingInfo): Boolean {
        val keyPair = client.currentKeyPair ?: throw NoActiveKeyPairException()
        return client.core.pollForAuthorization(
            info.server,
            info.digest_sha256.toByteArray(),
            keyPair.toPublicKeyProto().toFfi(),
        )
    }

    private suspend fun signIssuerState(issuerState: IssuerPairingState): SignedIssuerState {
        val keyPair = client.currentKeyPair ?: throw NoActiveKeyPairException()
        val stateBytes = IssuerPairingState.ADAPTER.encode(issuerState)
        val signature = client.crypto.sign(keyPair.privateKey, stateBytes, keyPair.keyType)
        return SignedIssuerState(
            state_bytes = stateBytes.toByteString(),
            signature = signature.toByteString(),
        )
    }

    /** Sign and publish new session state derived from the local identity chain. */
    private suspend fun putState(server: String, digestBytes: ByteArray, sequence: Long): PairingSession {
        val digest = digestFrom(digestBytes, assertIssuer = true)
        val identityState = client.listValidEvents(digest.issuer_identity, Collections.IDENTITY)
            .lastOrNull()
            ?: throw PolycentricException("No local identity chain for ${digest.issuer_identity}")

        val signedState = signIssuerState(
            IssuerPairingState(
                session_digest = digestBytes.toByteString(),
                identity_state = identityState,
                sequence = sequence,
            ),
        )

        val responseBytes = client.core.putPairingSession(
            server,
            SignedIssuerState.ADAPTER.encode(signedState),
        )
        return decodeSession(server, responseBytes)
    }

    /**
     * Decode a `PairingSessionDigest`. With [assertIssuer], verify the
     * session belongs to the active identity and key pair.
     */
    private fun digestFrom(digestBytes: ByteArray, assertIssuer: Boolean = false): PairingSessionDigest {
        val decoded = PairingSessionDigest.ADAPTER.decode(digestBytes)
        val signer = decoded.issuer_signer
            ?: throw PolycentricException("Pairing session digest has no issuer signer")
        val digest = decoded.copy(issuer_signer = signer)

        if (assertIssuer) {
            if (digest.issuer_identity != client.activeIdentityKey) {
                throw PolycentricException(
                    "Pairing session specifies a different identity than the active identity key",
                )
            }
            val keyPair = client.currentKeyPair ?: throw NoActiveKeyPairException()
            if (!IdentityManager.keysEqual(signer, keyPair.toPublicKeyProto())) {
                throw PolycentricException(
                    "Pairing session pins a different signer than the active key pair",
                )
            }
        }

        return digest
    }

    /** Decode a `PairingSessionState` received from a server. */
    private fun decodeSession(server: String, stateBytes: ByteArray): PairingSession {
        val state = PairingSessionState.ADAPTER.decode(stateBytes)
        val signedIssuerState = state.issuer_state
            ?: throw PolycentricException("Pairing session state has no signed issuer state")
        val issuerState = IssuerPairingState.ADAPTER.decode(signedIssuerState.state_bytes.toByteArray())
        val digestBytes = issuerState.session_digest.toByteArray()
        val digest = digestFrom(digestBytes)
        val expiresAt = digest.initial_timestamp + digest.ttl_millis
        if (expiresAt < digest.initial_timestamp) {
            throw PolycentricException("Pairing session expiration overflows")
        }

        return PairingSession(
            digestBytes = digestBytes,
            digest = digest,
            issuerState = issuerState,
            claimers = state.claimers,
            expiresAt = expiresAt,
            pairingInfo = PairingInfo(
                server = server,
                digest_sha256 = sha256(digestBytes).toByteString(),
            ),
        )
    }

    private fun randomNonce(): ByteString {
        val nonce = ByteArray(NONCE_BYTES)
        SecureRandom().nextBytes(nonce)
        return nonce.toByteString()
    }

    private fun sha256(bytes: ByteArray): ByteArray =
        MessageDigest.getInstance("SHA-256").digest(bytes)

    private fun FfiPublicKey.toProto(): PublicKey = PublicKey(
        key_type = KeyType.fromValue(keyType) ?: KeyType.KEY_TYPE_UNSPECIFIED,
        key = key.toByteString(),
    )
}
